### PR TITLE
feat: define region server and related requests

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR defines the `RegionServer` struct, `RegionEngine` trait, the method to handle requests and related requests. 

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/store-api/src/region_request.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    Compact(RegionCompactRequest),
+}
+
+/// Request to write a region.
+#[derive(Debug)]
+pub struct RegionWriteRequest {
+    /// Rows to write.
+    pub rows: Rows,
+    /// Map column name to column index in `rows`.
+    pub name_to_index: HashMap<String, usize>,
```
Comment (evenyag) [inline src/store-api/src/region_request.rs:42]: Better to make name_to_index private and generated in `new` to ensure it is correct.

File: src/datanode/src/region_server.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                    .get(&region_id)
+                    .with_context(|| RegionNotFoundSnafu { region_id })?;
+                _dashmap_guard = Some(guard);
+                _dashmap_guard.as_ref().map(|g| &*g).unwrap()
+            }
+        };
+
+        let engine =
+            self.engines
+                .get(&*engine_type)
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/datanode/src/region_server.rs:81]: Is it good to hold a ref for each region id....?

File: src/store-api/src/region_request.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use datatypes::arrow::row::Rows;
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/store-api/src/region_request.rs:17]: Looks like it's not merged yet?

File: src/store-api/src/region_request.rs (hunk 1 above)
Comment (waynexia) [reply src/store-api/src/region_request.rs]: Remove it for now. We can add it back later when implement utils for those utils, if necessary

File: src/mito2/src/worker/handle_create.rs [hunk 4]
```diff
@@ -18,44 +18,54 @@ use std::sync::Arc;
...
-        let mut builder = RegionMetadataBuilder::new(request.region_id, INIT_REGION_VERSION);
+        let mut builder = RegionMetadataBuilder::new(region_id, INIT_REGION_VERSION);
         for column in request.column_metadatas {
             builder.push_column_metadata(column);
         }
         builder.primary_key(request.primary_key);
         let metadata = builder.build()?;
 
+        // TODO(yingwen): Is it necessary to put this in the request option?
+        let region_dir = request
```
Comment (evenyag) [inline src/mito2/src/worker/handle_create.rs:55]: We also need the catalog/schema/table id to get the correct dir.

**Review by v0y4g3r — APPROVED** [review]
Test approve

**Review by evenyag — APPROVED** [review]

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**